src/shogun/classifier/AveragedPerceptron.cpp

```diff
@@ -14,105 +15,99 @@
 
 using namespace shogun;
 
-CAveragedPerceptron::CAveragedPerceptron()
-: CLinearMachine()
+CAveragedPerceptron::CAveragedPerceptron() :
+ 	CAveragedPerceptron(nullptr, nullptr, 1000, 0.1)
 {
-	init();
-
 }
 
-CAveragedPerceptron::CAveragedPerceptron(CDotFeatures* traindat, CLabels* trainlab)
-: CLinearMachine()
+CAveragedPerceptron::CAveragedPerceptron(CDotFeatures* traindat, CLabels* trainlab) :
+	CAveragedPerceptron(traindat, trainlab, 1000, 0.1)
 {
-	set_features(traindat);
-	set_labels(trainlab);
-	init();
 }
 
-CAveragedPerceptron::~CAveragedPerceptron()
+CAveragedPerceptron::CAveragedPerceptron(
+	CDotFeatures* traindat,
+	CLabels* trainlab,
+	int32_t max_iters,
+	float64_t learning_rate
+) : CIterativeMachine<CLinearMachine>()
 {
+	if (trainlab) set_labels(trainlab);
+	if (traindat) init_model((CFeatures*) traindat);
+	
+	set_max_iterations(max_iters);
+	set_learn_rate(learning_rate);
+	
+	SG_ADD(&learn_rate, "learn_rate", "Learning rate.", ParameterProperties::HYPER);
```

Comment:
This needs to happen in every ctor. Most importantly the default one. So best to move those to a method